### PR TITLE
EntityRef/Mut get_components

### PR DESCRIPTION
**Pull request by cart**
# Objective

Implements #13127

`EntityRef` / `EntityMut` usage is currently unnecessarily verbose and borrow-checker-constrained. It is currently impossible to use `EntityMut` to get mutable references to more than one component at the same time:

```rust
let mut entity = world.entity_mut(some_entity);
let mut a = entity.get_mut::<A>();
// this fails to compile because `entity` is already borrowed mutably
let mut b = entity.get_mut::<B>();
```

This makes `EntityMut` significantly less useful than it could be. Querying necessary components on demand is a pattern many prefer to use over traditional top-level queries. 

## Solution

Implement new `get_components` and `components` methods for `EntityRef` and `EntityMut` (and mutable variants for `EntityMut`) to provide multi-component access:

```rust
let (a, b) = world.entity(SOME_ENTITY).components::<(&A, &B)>();
let (a, b) = world.entity(SOME_ENTITY).get_components::<(&A, &B)>().unwrap();
let (mut a, b) = world.entity_mut(SOME_ENTITY).components_mut::<(&mut A, &B)>();
let (mut a, b) = world.entity_mut(SOME_ENTITY).get_components_mut::<(&mut A, &B)>().unwrap();
```

This is implemented using our existing Query implementation to provide multi-component access. It checks access without actually computing QueryState / access using the very convenient `Q::matches_component_set`.

## Next Steps

* This doesn't add `FilteredEntityRef` and `FilteredEntityMut` variants, as checking that access is actually more challenging (and with the current methods, likely prohibitively expensive because I think we need to compute `Access<ComponentId>` )
* This doesn't do the `components` to `get` and `get_components` to `try_get` renaming / unification I proposed in #13127 as I suspect that will be more controversial, and it also touches a lot of code. 

---

### Added

- `EntityRef::components`, `EntityRef::get_components`, `EntityMut::components`, `EntityMut::get_components`, `EntityMut::components_mut`, and `EntityMut::get_components_mut`.

### Changed

- Renamed `FilteredEntityRef::components` to `FilteredEntityRef::accessed_components` and `FilteredEntityMut::components` to `FilteredEntityMut::accessed_components` to avoid future name clashes

## Migration Guide

- Rename `FilteredEntityRef::components` to `FilteredEntityRef::accessed_components` and `FilteredEntityMut::components` to `FilteredEntityMut::accessed_components`

### CODE REVIEW

File: crates/bevy_ecs/src/world/entity_ref.rs [hunk 1]
```diff
@@ -153,6 +153,17 @@ impl<'w> EntityRef<'w> {
...
     }
+
+    /// Returns read-only components for the current entity that match the query `Q`.
+    pub fn components<Q: ReadOnlyQueryData>(&self) -> Q::Item<'_> {
+        self.get_components::<Q>().unwrap()
+    }
+
+    /// Returns read-only components for the current entity that match the query `Q`.
+    pub fn get_components<Q: ReadOnlyQueryData>(&self) -> Option<Q::Item<'_>> {
+        // SAFETY: &mut self implies exclusive access for duration of returned value
```
Comment (cBournhonesque) [inline crates/bevy_ecs/src/world/entity_ref.rs:164]:
```suggestion
        // SAFETY: We have read-only access to all components of this entity.
```

**Comment by msvbg** [comment]
In light of #12660, I would vote against adding more panicking shorthands like `components`.

File: crates/bevy_ecs/src/world/entity_ref.rs [hunk 2]
```diff
@@ -348,6 +359,28 @@ impl<'w> EntityMut<'w> {
...
+    }
+
+    /// Returns components for the current entity that match the query `Q`.
+    pub fn components_mut<Q: QueryData>(&mut self) -> Q::Item<'_> {
+        self.get_components_mut::<Q>().unwrap()
+    }
+
+    /// Returns read-only components for the current entity that match the query `Q`.
+    pub fn get_components<Q: ReadOnlyQueryData>(&self) -> Option<Q::Item<'_>> {
+        // SAFETY: &mut self implies exclusive access for duration of returned value
```
Comment (cBournhonesque) [inline crates/bevy_ecs/src/world/entity_ref.rs:374]:
```suggestion
        // SAFETY: We have read-only access to all components of this entity.
```

**Review by cBournhonesque — APPROVED** [review]
Would this also allow for
```
let (mut a, mut b) = world.entity_mut(SOME_ENTITY).components_mut::<(&mut A, &mut B)>();
```
?

Maybe we can add a unit test for that case as well

**Comment by cart** [comment]
> In light of https://github.com/bevyengine/bevy/issues/12660, I would vote against adding more panicking shorthands like components.

I think discouraging unwrap/panicking variants entirely (as in, not providing impls) would have pretty nasty ergonomics implications for people, especially with the current Option vs Result compatibility problem for `?`. I would be very uncomfortable removing those apis right now.

**Comment by ramirezmike** [comment]
> > In light of #12660, I would vote against adding more panicking shorthands like components.
> 
> I think discouraging unwrap/panicking variants entirely (as in, not providing impls) would have pretty nasty ergonomics implications for people, especially with the current Option vs Result compatibility problem for `?`. I would be very uncomfortable removing those apis right now.

I agree somewhat with this. It's frustrating when something like `despawn` causes a panic due to an unwrap, but if I want to shoot myself in the foot with something like `single` _or_ `components_mut` I should be allowed to.

**Comment by cart** [comment]
Left some replies in #12660

File: crates/bevy_ecs/src/world/entity_ref.rs [hunk 3]
```diff
@@ -2918,4 +2978,24 @@ mod tests {
...
+        let e2 = world.spawn(X(8)).id();
+        let e3 = world.spawn_empty().id();
+
+        assert_eq!(
+            Some((&X(7), &Y(10))),
+            world.entity(e1).get_components::<(&X, &Y)>()
+        );
+        assert_eq!(None, world.entity(e2).get_components::<(&X, &Y)>());
+        assert_eq!(None, world.entity(e3).get_components::<(&X, &Y)>());
+    }
```
Comment (hymm) [inline crates/bevy_ecs/src/world/entity_ref.rs:3000]: can you add a test to make sure it returns none or panics if you do `world.entity_mut(e).get_components_mut::<&mut X, &mut X>()`?

File: crates/bevy_ecs/src/world/entity_ref.rs [hunk 4]
```diff
@@ -153,6 +153,17 @@ impl<'w> EntityRef<'w> {
...
         unsafe { self.0.get_by_id(component_id) }
     }
+
+    /// Returns read-only components for the current entity that match the query `Q`.
+    pub fn components<Q: ReadOnlyQueryData>(&self) -> Q::Item<'w> {
+        self.get_components::<Q>().expect(QUERY_MISMATCH_ERROR)
+    }
+
+    /// Returns read-only components for the current entity that match the query `Q`.
+    pub fn get_components<Q: ReadOnlyQueryData>(&self) -> Option<Q::Item<'w>> {
```
Comment (hymm) [inline crates/bevy_ecs/src/world/entity_ref.rs:163]: Should add some docs to explain why it might return `None`. Ditto for the other methods that return an option.

File: crates/bevy_ecs/src/world/unsafe_world_cell.rs [hunk 5]
```diff
@@ -818,6 +819,48 @@ impl<'w> UnsafeEntityCell<'w> {
...
+        let state = Q::get_state(self.world().components())?;
+        let location = self.location();
+        // SAFETY: Location is guaranteed to exist
+        let archetype = unsafe {
+            self.world
+                .archetypes()
+                .get(location.archetype_id)
+                .debug_checked_unwrap()
+        };
+        if Q::matches_component_set(&state, &|id| archetype.contains(id)) {
```
Comment (james7132) [inline crates/bevy_ecs/src/world/unsafe_world_cell.rs:837]: Could use `bool::then` instead of the if/else here.

**Review by james7132 — COMMENTED** [review]
I find this kind of funny. Thinking about this historically, this feels like we've inverted the relationship between `Query::get_component` and `EntityMut`/`EntityRef` with this change.

>I think discouraging unwrap/panicking variants entirely (as in, not providing impls) would have pretty nasty ergonomics implications for people, especially with the current Option vs Result compatibility problem for ?. I would be very uncomfortable removing those apis right now.

I agree that we probably shouldn't remove them until `try`-blocks or some equivalent are stabilized.

**Comment by ramirezmike** [comment]
Since this uses `QueryData`, this would allow doing stuff like this, right?

```world.entity(e).get_components::<(Entity, &X, Option<&Y>, Has<Z>)>();```

"components" sorta makes sense for Option/Has... Is Entity a component though? What if QueryData gets implemented with more Component-adjacent stuff in the future?

I'm guessing the next steps item 2 would be renaming these to the `get` functions in the future?

File: crates/bevy_ecs/src/world/entity_ref.rs (hunk 3 above)
Comment (hymm) [reply crates/bevy_ecs/src/world/entity_ref.rs:3000]: also should have one for `(&mut X, &X)`

**Comment by cart** [comment]
> I find this kind of funny. Thinking about this historically, this feels like we've inverted the relationship between Query::get_component and EntityMut/EntityRef with this change.

Haha yeah thats an interesting comparison. Although I think this is significantly more defensible / necessary than `Query::get_component`.

> Since this uses QueryData, this would allow doing stuff like this, right?

Yup!

> I'm guessing the next steps item 2 would be renaming these to the get functions in the future?

Yup that is something discussed in the Next Steps section above. Haven't thought of a better interim name. And `components` is _approximately_ correct.

**Comment by cart** [comment]
Welp this PR is sadly stalled until we sort out this problem:
https://github.com/bevyengine/bevy/pull/13375#discussion_r1602135098

Not sure there will be a solution that doesn't involve doing access checks on each call.

**Comment by cart** [comment]
I think the best we can hope for is "allocation free check every entry against every other entry" (similar to `query.many_mut`), but accomplishing that will require significant changes (or at least, additions) to how we compute access.

File: crates/bevy_ecs/src/world/entity_ref.rs (hunk 3 above)
Comment (cBournhonesque) [reply crates/bevy_ecs/src/world/entity_ref.rs:3000]: Wouldn't creating an empty access and calling `update_component_access(state, &mut access)` on it work?

Comment (cart) [reply crates/bevy_ecs/src/world/entity_ref.rs:3000]: It would. It would just be dissatisfying from a perf perspective because Access allocates.

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
I'm in favor of this existing! I agree with the request for more thorough docs (shock). Nonblocking, but I also think we should be returning `Result`, not `Option` here (as with most of our ECS APIs). [QueryEntityError](https://dev-docs.bevyengine.org/bevy/ecs/query/enum.QueryEntityError.html) looks like the closest fit, but would need to be tweaked to better fit the actual problems encountered (and API used).

Obviously the UB is a blocker here on this getting merged. Given the discovered complexity, I'm removing this from the 0.14 milestone <3 I still want this, and if it makes it it makes it, but it's not correct to try and wait for this to be resolved.

**Comment by alice-i-cecile** [comment]
I think we should be able to reuse the access mechanisms from queries to validate soundness here. Would need to sit down with the code to figure out exactly how. I much prefer that over creating another mechanism though, especially since I'd like to be able to add untyped variants to this in the future.

File: crates/bevy_ecs/src/world/entity_ref.rs (hunk 3 above)
Comment (cBournhonesque) [reply crates/bevy_ecs/src/world/entity_ref.rs:3000]: FYI I have a PR that removes the allocation from an empty `Access`: https://github.com/bevyengine/bevy/pull/14026/files which might unblock this?

Comment (iiYese) [reply crates/bevy_ecs/src/world/entity_ref.rs:3000]: Could the allocation not just be removed in the meantime by using a smallvec that has 8 items or something?

**Comment by ItsDoot** [comment]
Would it be possible to merge a version of this PR with just the immutable `get_components` variants for `0.15`? Even just the immutable variants would be very useful for me.

**Comment by alice-i-cecile** [comment]
I agree the an immutable form would be useful still. Feel free to spin off a PR that does that please.

**Comment by alice-i-cecile** [comment]
Closing in favor of #15089; further attempts to implement the mutable form will be easier in a separate PR.

**Comment by alice-i-cecile** [comment]
Ultimately fixed by #21780.
